### DISCUSSION TITLE
Is it possible to tweak programming timing somehow?

### DISCUSSION THREAD

**Discussion by kekopop** [TomNisbet/TommyPROM#47] [Q&A]
Trying to erase a few Xicor X2864AP-45 EEPROMs by filling them with FF, without success. Some bytes will not erase properly. There's a certain pattern to it, if relevant somehow:

![tommyprom](https://user-images.githubusercontent.com/103499578/211681868-b69adfd6-6e0d-48a0-ae07-91356b5b7ced.jpg)

(Only dumped the last 512 bytes to keep the screenshot reasonably small, but it looks more or less the same everywhere) 

The programmer works perfectly with other 64kbit 28 series EEPROMs. 

Those chips might be ancient, but they tend to be more or less immortal. They haven't been stored under conditions outside what's specified in the datasheet. With other words I don't believe there's anything wrong with the chips either.

The X2864AP-45 is however a 450 ns chip. Can't even remember seeing any 28 series chip that slow before. Could it perhaps be some kind of timing issue, that the programmer is designed with faster chips in mind, and simply goes too fast?

Had a look at the code, but couldn't see any obvious ways to tweak the timing. Just wondering how to go about it?

Also, as a suggestion for future releases, would be nice to be able to tweak the timing from configure.h or something :)

**Comment by TomNisbet** [comment]
That chip has a 16 byte buffer for block writes.  Did you change that at the top of TommyPROM.ino where the chip is declared?  Your device declaration line should look like this:
`PromDevice28C  prom(8 * 1024L, 16, 10, true);`
You could also change the 16 to a zero to disable block writes entirely and write the chip a byte at a time.  If that isn't the problem then we can slow the code down.  Let me know if it works so I can add it to the documentation.

**Reply by kekopop** [reply]
Thanks for addressing this so quickly :)

Block writes have been disabled from the beginning, declaration line `PromDevice28C prom(8 * 1024L, 0, 10, true);` Setting the block size to 16 bytes strangely enough makes the programmer error out immediately, with a generic message, "ERROR: Write failed".

Also tried going back and forth between 0 and 16 bytes a couple of times to verify the error wasn't caused by something else. No write failures with block size reset to 0, nothing changed otherwise.

So, lets try slowing the code down, if you don't have anything else you want me to try before :)

**Comment by TomNisbet** [comment]
That is confusing. With a block size of zero, the code will be writing a single byte and then reading it back to verify.  The fill command should have failed above at 1c0f and printed a message.  Can you try using the zap command and posting the result?  I'd like to see what happens when the data between each byte is different.  Also, there is an Info "I" command that will print some additional debug information after a failure.

I looked at the data sheet and the shortest times between the signals are 450ns.  The current code is delaying for 1000ns between the signals, so that *shouldn't* be the problem.  If you want to increase the delays, look at burnByte and waitForWriteCycleEnd in the PromDevice28C.cpp file.  You can change the delays and add new ones.  Try adding a delay after setAddress() in burnByte.

**Comment by kekopop** [comment]
Okay, so the programmer verifies written data byte by byte. Yes, very confusing then. If we're assuming the code is doing what it's supposed to, the only reasonable explanation would be that some memory cells loses charge or something, and the data changes between writing/verifying and me dumping the memory contents to console. I thought those old EEPROMs were more or less immortal, but might have to change my opinion on that then.

Tried filling the chip with 00, then filling it with FF, and started checksumming it periodically immediately afterwards. The data looks stable enough within that timeframe at least:

```
>f0 1fff 00

>f0 1fff ff

>c0 1fff

Checksum 0000-1fff = ff82

>c0 1fff

Checksum 0000-1fff = ff82

>c0 1fff

Checksum 0000-1fff = ff82

>c0 1fff

Checksum 0000-1fff = ff82

>c0 1fff

Checksum 0000-1fff = ff82
```
There's of course no way for me to check what happens the micro- or milliseconds directly after writing to the cells though...

----

The zap command fails at 0xF, consistent with how it looks after a fill. Block writes disabled if that matters somehow. It's always byte E or F in each row that's bad. With zap, it errors out as expected. Are the zap and fill verification routines perhaps different?

```
>z0

ERROR: Verify failed  addr=0xF  read=0xFF  expected=0x80

>d0 1fff
00000: 41 42 43 44  45 46 47 48  01 02 04 08  10 20 40 ff  ABCDEFGH ..... @.
00010: 7f bf df ef  f7 fb fd fe  00 ff 55 aa  30 31 32 33  ........ ..U.0123
00020: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00030: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
00040: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
```

Enabled 16 byte block writes, tried the fill command, and printed debug info:
```
>f0 1fff

ERROR: Write failed

>i
debugBlockWrites:  1
debugLastAddress:  F
debugLastExpected: 0
debugLastReadback: FF
debugRxDuplicates: 0
debugExtraChars:   0
debugRxStarts:     0
debugRxSyncErrors: 0
```

I'm very much a newbie when it comes to programming, but I'll try figuring out how to manipulate the delays, and revert with the results in a separate post. I'm also thinking it might be worth trying with longer programming pulses aswell, if you can help me with that too. Big thanks for your help so far in any case 😄

**Comment by kekopop** [comment]
Raised all 1 μs delays found in burnByte and waitForWriteCycleEnd - five of them in total -  to 2 μs, and believe it or not, it works perfectly now 😄 

Tried filling with different byte patterns, erasing, and zapping a few randomly chosen address spaces. No errors at all anymore.

Also found the burnBlock function. Tried raising the delays in it to 2 μs aswell, and enable 16 byte block writing, but that still throws me "ERROR: Write failed". I'm happy without block writes, but if you want to try figuring it out, I'm of course at your service with testing.

Also, it's really strange that it passed the byte by byte verification before, please tell me if I can help you with any further testing of that too.

If it's possible to ship from Sweden to where you are in a regular letter, avoiding customs declaration and associated fees/administration, etc, I could also send you a chip to experiment with if you want.

**Comment by TomNisbet** [comment]
It would be interesting to see the results of a zap command to address 1 or 2 instead of zero.  Will the problem byte move or will it stay at the 16 byte boundary?  

If you have a spare chip to send to the US, I would like to try programming it while watching the signals on the logic analyzer.  Please send a DM to nib85 on reddit if that is something you are willing to do.

In any case, I'm glad the code is working for you.  What sort of project are you doing with those chips?

**Comment by kekopop** [comment]
Filled the chip with 00, reverted the delays to 1 μs, then filled it with FF. The first 128 bytes after that:

```
00000: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00010: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
00020: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
00030: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff 00 00  ........ ........
00040: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00050: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00060: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00070: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
00080: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00090: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
000a0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
000b0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
000c0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff 00  ........ ........
000d0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff 00 00  ........ ........
000e0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
000f0: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff 00 00  ........ ........
```

Did some zapping with different offsets:


```
>z11

INFO: Write test successful

>z32

INFO: Write test successful

>z73

INFO: Write test successful

>za4

ERROR: Verify failed  addr=0xBF  read=0x0  expected=0xAA

>zc5

INFO: Write test successful

>ze6

ERROR: Verify failed  addr=0xFF  read=0x0  expected=0xFF

>d0 ff
00000: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00010: ff 41 42 43  44 45 46 47  48 01 02 04  08 10 20 40  .ABCDEFG H..... @
00020: 80 7f bf df  ef f7 fb fd  fe 00 ff 55  aa 30 31 32  ........ ...U.012
00030: 33 ff 41 42  43 44 45 46  47 48 01 02  04 08 10 20  3.ABCDEF GH.....
00040: 40 80 7f bf  df ef f7 fb  fd fe 00 ff  55 aa 30 31  @....... ....U.01
00050: 32 33 ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  23...... ........
00060: ff ff ff ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  ........ ........
00070: ff ff ff 41  42 43 44 45  46 47 48 01  02 04 08 10  ...ABCDE FGH.....
00080: 20 40 80 7f  bf df ef f7  fb fd fe 00  ff 55 aa 30   @...... .....U.0
00090: 31 32 33 ff  ff ff ff ff  ff ff ff ff  ff ff ff ff  123..... ........
000a0: ff ff ff ff  41 42 43 44  45 46 47 48  01 02 04 08  ....ABCD EFGH....
000b0: 10 20 40 80  7f bf df ef  f7 fb fd fe  00 ff 55 00  . @..... ......U.
000c0: 30 31 32 33  00 41 42 43  44 45 46 47  48 01 02 04  0123.ABC DEFGH...
000d0: 08 10 20 40  80 7f bf df  ef f7 fb fd  fe 00 ff 55  .. @.... .......U
000e0: aa 30 31 32  33 ff 41 42  43 44 45 46  47 48 01 02  .0123.AB CDEFGH..
000f0: 04 08 10 20  40 80 7f bf  df ef f7 fb  fd fe 00 00  ... @... ........

```
Before, zapping would fail at adress 1F iirc, but if it is a timing issue and we're close to the limit, I guess some randomness is to be expected. Both times it fails, it's on the 16 byte boundary.

I'm not at Reddit, but improvised and created a private repository here with my mail address in the description and invited you to it instead. Please mail me your details, and I'll send you a chip :) Github should really add a PM feature btw...

Salvaged the chips from a decommissioned 80's marine automation system (ABB Selma), planning to use at least some of them for custom cartridges and other C64 stuff :)
